core/arch/arm/mm/mobj_ffa.c

```diff
@@ -0,0 +1,562 @@
+{
+	return mf == (void *)(vaddr_t)ptr;
+}
+
+static struct mobj_ffa *pop_from_list(struct mobj_ffa_head *head,
+				      bool (*cmp_func)(struct mobj_ffa *mf,
+						       uint64_t val),
+				      uint64_t val)
+{
+	struct mobj_ffa *mf = SLIST_FIRST(head);
+	struct mobj_ffa *p = NULL;
+
+	if (!mf)
+		return NULL;
+
+	if (cmp_func(mf, val)) {
+		SLIST_REMOVE_HEAD(head, link);
+		return mf;
+	}
+
+	while (true) {
+		p = SLIST_NEXT(mf, link);
+		if (!p)
+			return NULL;
+		if (cmp_func(p, val)) {
+			SLIST_REMOVE_AFTER(mf, link);
+			return p;
+		}
+		mf = p;
+	}
+}
+
+static struct mobj_ffa *find_in_list(struct mobj_ffa_head *head,
+				     bool (*cmp_func)(struct mobj_ffa *mf,
+						      uint64_t val),
+				     uint64_t val)
+{
+	struct mobj_ffa *mf = NULL;
+
+	SLIST_FOREACH(mf, head, link)
+		if (cmp_func(mf, val))
+			return mf;
+
+	return NULL;
+}
+
+#ifdef CFG_CORE_SEL1_SPMC
+void mobj_ffa_sel1_spmc_delete(struct mobj_ffa *mf)
+{
+	int i = (mf->cookie - 1) & ~BIT64(44);
+	uint32_t exceptions = 0;
+
+	assert(i >= 0 && i < NUM_SHMS);
+
+	exceptions = cpu_spin_lock_xsave(&shm_lock);
+	assert(bit_test(shm_bits, i));
+	bit_clear(shm_bits, i);
+	assert(!mf->mm);
+	cpu_spin_unlock_xrestore(&shm_lock, exceptions);
+
+	free(mf);
+}
+#endif /*CFG_CORE_SEL1_SPMC*/
+
+TEE_Result mobj_ffa_add_pages_at(struct mobj_ffa *mf, unsigned int *idx,
+				 paddr_t pa, unsigned int num_pages)
+{
+	unsigned int n;
+	size_t tot_page_count = get_page_count(mf);
+
+	if (ADD_OVERFLOW(*idx, num_pages, &n) || n > tot_page_count)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!core_pbuf_is(CORE_MEM_NON_SEC, pa, num_pages * SMALL_PAGE_SIZE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	for (n = 0; n < num_pages; n++)
+		mf->pages[n + *idx] = pa + n * SMALL_PAGE_SIZE;
+
+	(*idx) += n;
+	return TEE_SUCCESS;
+}
+
+uint64_t mobj_ffa_get_cookie(struct mobj_ffa *mf)
+{
+	return mf->cookie;
+}
+
+uint64_t mobj_ffa_push_to_inactive(struct mobj_ffa *mf)
+{
+	uint32_t exceptions = 0;
+
+	exceptions = cpu_spin_lock_xsave(&shm_lock);
+	assert(!find_in_list(&shm_inactive_head, cmp_ptr, (vaddr_t)mf));
+	assert(!find_in_list(&shm_inactive_head, cmp_cookie, mf->cookie));
+	assert(!find_in_list(&shm_head, cmp_cookie, mf->cookie));
+	SLIST_INSERT_HEAD(&shm_inactive_head, mf, link);
+	cpu_spin_unlock_xrestore(&shm_lock, exceptions);
+
+	return mf->cookie;
+}
+
+static void unmap_helper(struct mobj_ffa *mf)
+{
+	if (mf->mm) {
+		core_mmu_unmap_pages(tee_mm_get_smem(mf->mm),
+				     get_page_count(mf));
+		tee_mm_free(mf->mm);
+		mf->mm = NULL;
+	}
+}
+
+TEE_Result mobj_ffa_unregister_by_cookie(uint64_t cookie)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct mobj_ffa *mf = NULL;
+	uint32_t exceptions = 0;
+
+	exceptions = cpu_spin_lock_xsave(&shm_lock);
+	mf = find_in_list(&shm_head, cmp_cookie, cookie);
+	/*
+	 * If the mobj is found here it's still active and cannot be
+	 * unregistered.
+	 */
+	if (mf) {
+		DMSG("cookie %#"PRIx64" busy refc %u",
```

Comment:
suggest to move the trace out of the spinlock xsave.

Reply:
Why? What's the concern?

Reply:
I though about evacuating trace with interrupt masked.
Ok as is.